spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -538,68 +543,128 @@ protected WriterFactory(Broadcast<Table> tableBroadcast, FileFormat format, long
     @Override
     public DataWriter<InternalRow> createWriter(int partitionId, long taskId, long epochId) {
       Table table = tableBroadcast.value();
-
-      OutputFileFactory fileFactory = OutputFileFactory.builderFor(table, partitionId, taskId).format(format).build();
-      SparkAppenderFactory appenderFactory = SparkAppenderFactory.builderFor(table, writeSchema, dsSchema).build();
-
       PartitionSpec spec = table.spec();
       FileIO io = table.io();
 
+      OutputFileFactory fileFactory = OutputFileFactory.builderFor(table, partitionId, taskId)
+          .format(format)
+          .build();
+      SparkFileWriterFactory writerFactory = SparkFileWriterFactory.builderFor(table)
+          .dataFileFormat(format)
+          .dataSchema(writeSchema)
+          .dataSparkType(dsSchema)
+          .build();
+
       if (spec.isUnpartitioned()) {
-        return new Unpartitioned3Writer(spec, format, appenderFactory, fileFactory, io, targetFileSize);
+        ClusteredDataWriter<InternalRow> dataWriter = new ClusteredDataWriter<>(
+            writerFactory, fileFactory, io,
+            format, targetFileSize);
+        return new UnpartitionedDataWriter(dataWriter, io, spec);
+
       } else if (partitionedFanoutEnabled) {
-        return new PartitionedFanout3Writer(
-            spec, format, appenderFactory, fileFactory, io, targetFileSize, writeSchema, dsSchema);
+        FanoutDataWriter<InternalRow> dataWriter = new FanoutDataWriter<>(
+            writerFactory, fileFactory, io,
+            format, targetFileSize);
+        return new PartitionedDataWriter(dataWriter, io, spec, writeSchema, dsSchema);
+
       } else {
-        return new Partitioned3Writer(
-            spec, format, appenderFactory, fileFactory, io, targetFileSize, writeSchema, dsSchema);
+        ClusteredDataWriter<InternalRow> dataWriter = new ClusteredDataWriter<>(
+            writerFactory, fileFactory, io,
+            format, targetFileSize);
+        return new PartitionedDataWriter(dataWriter, io, spec, writeSchema, dsSchema);
       }
     }
   }
 
-  private static class Unpartitioned3Writer extends UnpartitionedWriter<InternalRow>
-      implements DataWriter<InternalRow> {
-    Unpartitioned3Writer(PartitionSpec spec, FileFormat format, SparkAppenderFactory appenderFactory,
-                         OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
-      super(spec, format, appenderFactory, fileFactory, io, targetFileSize);
+  // TODO: why the old implementation throws throwFailureWhenFinished()?
```

Comment:
This is an open point. Any particular reasons for using `throwFailureWhenFinished` in the old implementation? Is it to better propagate any exceptions while deleting files?

Reply:
It seems like we do not have consistency of this across implementations. We did `throwFailureWhenFinished` for all the aborts in Spark, but for most of the actions we did `suppressFailureWhenFinished`. Overall my thoughts are the following:
1. I would prefer we keep the old behavior if possible because it's debatable which one is more desirable, so it's probably better to just keep the existing behavior
2. if we think suppression is better for some important reasons, we should also log a warning `onFailure` so that at least user has a way to know what is going on
2. ideally we can have a flag `throwOnAbortFailure` to control this from the user side.

Reply:
I also think it is safer to keep the old behavior.

I took a look and see that Spark uses `tryWithSafeFinallyAndFailureCallbacks` and will log the delete exception. I'll update to match the old behavior.

Reply:
Is the thought that we may have other operations that need to occur after this part of the code? For example, say we end up finishing up our Spark Staging Catalog code and actually code in an "on abort" there. If we fail here we would never get back to that Staging portion of the code and attempt the other cleanup.

Reply:
I think it is purely a data writer abort, which is responsible for writing data for a single task and it looks like Spark handles all exceptions gracefully for such cases.

If I were to implement this from scratch, I'd probably go for a warning and suppression. However, I am also worried about changing the behavior that worked so far :)